src/hotspot/share/opto/vectornode.cpp

```diff
@@ -1619,17 +1619,9 @@ Node* VectorUnboxNode::Ideal(PhaseGVN* phase, bool can_reshape) {
         bool is_vector_mask    = vbox_klass->is_subclass_of(ciEnv::current()->vector_VectorMask_klass());
         bool is_vector_shuffle = vbox_klass->is_subclass_of(ciEnv::current()->vector_VectorShuffle_klass());
         if (is_vector_mask) {
+          // VectorUnbox (VectorBox vmask) ==> VectorMaskCast vmask
           const TypeVect* vmask_type = TypeVect::makemask(out_vt->element_basic_type(), out_vt->length());
-          if (in_vt->length_in_bytes() == out_vt->length_in_bytes() &&
-              Matcher::match_rule_supported_vector(Op_VectorMaskCast, out_vt->length(), out_vt->element_basic_type())) {
-            // Apply "VectorUnbox (VectorBox vmask) ==> VectorMaskCast (vmask)"
-            // directly. This could avoid the transformation ordering issue from
-            // "VectorStoreMask (VectorLoadMask vmask) => vmask".
-            return new VectorMaskCastNode(value, vmask_type);
```

Comment:
Why do you change this code?
Is it a must to enable this optimization?

Reply:
For a vector mask unbox operation, if the input is a `VectorBox`, and the boxed value type is not equal to the unbox type, we need to do the type casting. Originally, we separate the type casting for two different cases:
1. if the vector element size in bytes is equal (e.g. casting between `int` and `float`), the conversion is:
```
  VectorUnbox (VectorBox vmask) ==> VectorMaskCast (vmask)
```
This doesn't need to emit any instructions.

2. otherwise (e.g. casting from `short` to `int`), the conversion is:
```
  VectorUnbox (VectorBox vmask) ==> VectorLoadMask (VectorStoreMask vmask)
```
This means we need to convert the short mask vector back to a boolean vector, and then casting the boolean vector to a int vector (i.e. `S->B->I`). 

With the new changes in this PR, `VectorMaskCast` supports the vector mask casting for all cases and all platforms (originally it only supports the first case for most platforms), we can implement the second cases simply with `VectorMaskCast`. That means we can optimize `S->B->I` to `S->I`. This saves the unnecessary narrowing and extending instructions due to the temporary boolean vector result.

Reply:
> For a vector mask unbox operation, if the input is a `VectorBox`, and the boxed value type is not equal to the unbox type, we need to do the type casting. Originally, we separate the type casting for two different cases:
> 
> 1. if the vector element size in bytes is equal (e.g. casting between `int` and `float`), the conversion is:
> 
> ```
>   VectorUnbox (VectorBox vmask) ==> VectorMaskCast (vmask)
> ```
> 
> This doesn't need to emit any instructions.
> 
> 2. otherwise (e.g. casting from `short` to `int`), the conversion is:
> 
> ```
>   VectorUnbox (VectorBox vmask) ==> VectorLoadMask (VectorStoreMask vmask)
> ```
> 
> This means we need to convert the short mask vector back to a boolean vector, and then casting the boolean vector to a int vector (i.e. `S->B->I`).
> 
> With the new changes in this PR, `VectorMaskCast` supports the vector mask casting for all cases and all platforms (originally it only supports the first case for most platforms), we can implement the second cases simply with `VectorMaskCast`. That means we can optimize `S->B->I` to `S->I`. This saves the unnecessary narrowing and extending instructions due to the temporary boolean vector result.



So there are at least two optimizations in this pr:
1) Opt for FIRST_NONZERO
2) Opt for vector mask unbox operation

How about splitting them and doing the second one in a new PR？

Reply:
Thanks for the advice! Yes, this patch firstly fix the performance issue for FIRST_NONZERO and together with some clean up for the related codes. I considered to split the changes before. However, to think more, if the patch only fixes the performance issue, the final codes will look mess and we need to add the additional complex IR checks for the vector mask cast (please see the initial fixing changes here https://openjdk.github.io/cr/?repo=jdk&pr=9737&range=00). Unifying the usage to `VectorMaskCast` IR makes the final code clean. So I prefer to make the whole changes together.

Reply:
> Thanks for the advice! Yes, this patch firstly fix the performance issue for FIRST_NONZERO and together with some clean up for the related codes. I considered to split the changes before. However, to think more, if the patch only fixes the performance issue, the final codes will look mess and we need to add the additional complex IR checks for the vector mask cast (please see the initial fixing changes here https://openjdk.github.io/cr/?repo=jdk&pr=9737&range=00). Unifying the usage to `VectorMaskCast` IR makes the final code clean. So I prefer to make the whole changes together.


I'm not sure if there would be some side effect about the vector unbox code change.

Since the original goal of this pr is optimizing `FIRST_NONZERO`, it might be better to do the others (code cleanup, unifying the usage of VectorMaskCast) in a separate pr to lower the risk.
Thanks.

Reply:
OK, I will split the patch based on your suggestion. Thanks again for the review!